File: src/models/digest.py
"""Digest data models."""

from datetime import datetime
from enum import StrEnum
from typing import TYPE_CHECKING

from pydantic import BaseModel, Field
from sqlalchemy import JSON, Boolean, Column, DateTime, Enum as SQLEnum, Integer, String, Text
from sqlalchemy.orm import Mapped, relationship

from src.models.base import Base
from src.models.query import ContentQuery

if TYPE_CHECKING:
    from src.models.audio_digest import AudioDigest


class DigestType(StrEnum):
    """Digest type."""

    DAILY = "daily"
    WEEKLY = "weekly"
    SUB_DIGEST = "sub_digest"  # Sub-digest for hierarchical digest creation


class DigestStatus(StrEnum):
    """Digest generation status."""

    PENDING = "PENDING"
    GENERATING = "GENERATING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    PENDING_REVIEW = "PENDING_REVIEW"  # NEW: Awaiting human review
    APPROVED = "APPROVED"  # NEW: Approved for delivery
    REJECTED = "REJECTED"  # NEW: Rejected (won't deliver)
    DELIVERED = "DELIVERED"


class Digest(Base):
    """Digest database model."""

    __tablename__ = "digests"

    id = Column(Integer, primary_key=True, autoincrement=True)
    digest_type = Column(SQLEnum(DigestType), nullable=False, index=True)

    # Time period covered
    period_start = Column(DateTime, nullable=False, index=True)
    period_end = Column(DateTime, nullable=False, index=True)

    # Content
    title = Column(String(500), nullable=False)
    executive_overview = Column(Text, nullable=False)
    strategic_insights = Column(JSON, nullable=False)  # List[Dict]
    technical_developments = Column(JSON, nullable=False)  # List[Dict]
    emerging_trends = Column(JSON, nullable=False)  # List[Dict]
    actionable_recommendations = Column(JSON, nullable=False)  # Dict[str, List[str]]
    sources = Column(JSON, nullable=False)  # List[Dict] with content references

    # Historical context
    historical_context = Column(JSON, nullable=True)  # List[Dict] from Graphiti

    # Unified content model fields (Phase 4)
    markdown_content = Column(Text, nullable=True)  # Full markdown representation
    theme_tags = Column(JSON, nullable=True)  # List[str] - Extracted theme tags
    source_content_ids = Column(JSON, nullable=True)  # List[int] - Content IDs used in digest

    # Metadata
    newsletter_count = Column(Integer, nullable=False)  # Legacy name, now represents content count
    status = Column(
        SQLEnum(DigestStatus),
        nullable=False,
        default=DigestStatus.PENDING,
        index=True,
    )
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow, index=True)
    completed_at = Column(DateTime, nullable=True)
    delivered_at = Column(DateTime, nullable=True)

    # Sharing
    is_public = Column(Boolean, nullable=False, default=False, server_default="false")
    share_token = Column(String(36), nullable=True, unique=True, index=True)

    # Generation metadata
    agent_framework = Column(String(100), nullable=False)
    model_used = Column(String(100), nullable=False)  # General model ID
    model_version = Column(String(20), nullable=True)  # Version
    token_usage = Column(Integer, nullable=True)
    processing_time_seconds = Column(Integer, nullable=True)

    # Review tracking
    reviewed_by = Column(String(200), nullable=True)
    review_notes = Column(Text, nullable=True)
    reviewed_at = Column(DateTime, nullable=True)
    revision_count = Column(Integer, default=0, nullable=False)
    revision_history = Column(JSON, nullable=True)  # Conversation audit trail

    # Hierarchical digest support (NEW)
    parent_digest_id = Column(Integer, nullable=True, index=True)  # FK added in migration
    child_digest_ids = Column(JSON, nullable=True)  # List[int] of child digest IDs
    is_combined = Column(Boolean, default=False, nullable=False)
    source_digest_count = Column(Integer, nullable=True)  # Number of sub-digests combined

    # Relationships
    audio_digests: Mapped[list["AudioDigest"]] = relationship(
        "AudioDigest",
        back_populates="digest",
        foreign_keys="AudioDigest.digest_id",
    )


class DigestSection(BaseModel):
    """A section within a digest (strategic insight, technical development, etc.)."""

    title: str = Field(..., description="Section title")
    summary: str = Field(..., description="2-3 sentence summary")
    details: list[str] = Field(default_factory=list, description="Detailed points")
    themes: list[str] = Field(default_factory=list, description="Related theme names")
    continuity: str | None = Field(None, description="Historical continuity text")
    followup_prompts: list[str] = Field(
        default_factory=list, description="LLM prompts for deeper analysis"
    )


class DigestData(BaseModel):
    """Pydantic model for digest data transfer."""

    digest_type: DigestType
    period_start: datetime
    period_end: datetime
    title: str
    executive_overview: str
    strategic_insights: list[DigestSection] = Field(default_factory=list)
    technical_developments: list[DigestSection] = Field(default_factory=list)
    emerging_trends: list[DigestSection] = Field(default_factory=list)
    actionable_recommendations: dict[str, list[str]] = Field(default_factory=dict)
    sources: list[dict] = Field(default_factory=list)
    historical_context: list[dict] | None = None

    # Unified content model fields
    markdown_content: str | None = None
    theme_tags: list[str] | None = None
    source_content_ids: list[int] | None = None

    newsletter_count: int
    agent_framework: str
    model_used: str  # General model ID
    model_version: str | None = None  # Version
    token_usage: int | None = None
    processing_time_seconds: float | None = None

    # Review tracking
    reviewed_by: str | None = None
    review_notes: str | None = None
    reviewed_at: datetime | None = None
    revision_count: int = 0
    revision_history: dict | None = None  # Conversation audit trail

    # Hierarchical digest support (NEW)
    parent_digest_id: int | None = None
    child_digest_ids: list[int] | None = None
    is_combined: bool = False
    source_digest_count: int | None = None


class DigestRequest(BaseModel):
    """Request parameters for digest generation."""

    digest_type: DigestType
    period_start: datetime
    period_end: datetime
    max_strategic_insights: int = Field(default=5, description="Max strategic insights")
    max_technical_developments: int = Field(default=5, description="Max technical developments")
    max_emerging_trends: int = Field(default=3, description="Max emerging trends")
    max_followup_prompts: int = Field(default=3, description="Max follow-up prompts per section")
    include_historical_context: bool = Field(default=True, description="Include historical context")
    content_query: "ContentQuery | None" = Field(
        default=None, description="Optional content selection override"
    )

File: web/src/types/digest.ts
/**
 * Digest Types
 *
 * TypeScript interfaces for digest entities.
 * These types mirror the backend Python models in src/models/digest.py
 *
 * Digests are the aggregated output combining multiple content summaries
 * into a cohesive document with multi-audience formatting.
 *
 * Field names use snake_case to match the Python backend.
 *
 * @see Backend model: src/models/digest.py
 */

/**
 * Type of digest
 * - daily: Aggregation of a single day's content
 * - weekly: Aggregation of a week's content
 * - sub_digest: Intermediate digest for hierarchical combination
 */
export type DigestType = "daily" | "weekly" | "sub_digest"

/**
 * Digest workflow status
 *
 * Tracks the digest through the review and delivery pipeline.
 */
export type DigestStatus =
  | "PENDING" // Created but not yet generated
  | "GENERATING" // LLM is generating content
  | "COMPLETED" // Generation complete, awaiting review
  | "FAILED" // Generation failed
  | "PENDING_REVIEW" // Ready for human review
  | "APPROVED" // Approved by reviewer
  | "REJECTED" // Rejected by reviewer
  | "DELIVERED" // Sent to recipients

/**
 * Source reference in a digest
 *
 * Links digest content back to original content items.
 */
export interface DigestSource {
  /** Newsletter title */
  title: string
  /** Publication name */
  publication: string | null
  /** Original date */
  date: string
  /** Original URL if available */
  url?: string | null
}

/**
 * Section within a digest
 *
 * Digests are structured into sections, each targeting
 * different aspects of the content.
 */
export interface DigestSection {
  /** Section title */
  title: string
  /** Brief summary of the section */
  summary: string
  /** Detailed points/bullets */
  details: string[]
  /** Related themes */
  themes: string[]
  /**
   * Historical continuity context
   * How this section relates to previous digests
   */
  continuity?: string | null
}

/**
 * Actionable recommendations organized by role
 *
 * Different recommendations for different audience segments.
 */
export interface ActionableRecommendations {
  /** For leadership */
  for_leadership?: string[]
  /** For teams */
  for_teams?: string[]
  /** For individuals */
  for_individuals?: string[]
  [key: string]: string[] | undefined
}

/**
 * Revision history entry
 *
 * Tracks changes made to the digest through the review process.
 */
export interface RevisionEntry {
  /** When the revision was made */
  timestamp: string
  /** Who made the revision */
  reviewer: string
  /** What action was taken */
  action: string
  /** Section-specific feedback */
  section_feedback?: Record<string, string>
}

/**
 * Digest list item (lightweight view)
 *
 * Used in list views - matches backend DigestSummary response model.
 */
export interface DigestListItem {
  id: number
  digest_type: DigestType
  title: string
  period_start: string
  period_end: string
  content_count: number
  status: DigestStatus
  created_at: string
  model_used: string
  revision_count: number
  reviewed_by: string | null
}

/**
 * Digest detail (full view)
 *
 * Used for detail views - matches backend DigestDetail response model.
 */
export interface DigestDetail {
  id: number
  digest_type: DigestType
  title: string
  period_start: string
  period_end: string
  executive_overview: string
  strategic_insights: DigestSection[]
  technical_developments: DigestSection[]
  emerging_trends: DigestSection[]
  actionable_recommendations: ActionableRecommendations
  sources: DigestSource[]
  content_count: number
  status: DigestStatus
  created_at: string
  completed_at: string | null
  model_used: string
  model_version: string | null
  processing_time_seconds: number | null
  revision_count: number
  reviewed_by: string | null
  reviewed_at: string | null
  review_notes: string | null
  is_combined: boolean
  child_digest_ids: number[] | null
}

/**
 * Digest statistics
 */
export interface DigestStatistics {
  total: number
  pending: number
  generating: number
  completed: number
  pending_review: number
  approved: number
  delivered: number
  by_type: Record<string, number>
}

/**
 * Digest entity (full model)
 *
 * The main aggregated output document combining
 * multiple newsletter summaries.
 */
export interface Digest extends DigestDetail {
  /** Historical context from knowledge graph */
  historical_context?: Record<string, unknown>
  /** Agent framework used */
  agent_framework: string
  /** Tokens used */
  token_usage: number | null
  /** Parent digest ID (for sub_digests) */
  parent_digest_id: number | null
  /** Number of source digests if combined */
  source_digest_count: number | null
  /** When delivered */
  delivered_at: string | null
  /** History of revisions */
  revision_history: RevisionEntry[] | null
}

/**
 * Request to generate a new digest
 */
export interface GenerateDigestRequest {
  /** Type of digest to generate */
  digest_type: "daily" | "weekly"
  /** Start of period to cover (optional, computed based on type) */
  period_start?: string
  /** End of period to cover (optional, defaults to now) */
  period_end?: string
  /** Max strategic insights */
  max_strategic_insights?: number
  /** Max technical developments */
  max_technical_developments?: number
  /** Max emerging trends */
  max_emerging_trends?: number
  /** Include historical context from knowledge graph */
  include_historical_context?: boolean
}

/**
 * Review action for a digest
 */
export type ReviewAction = "approve" | "request_revision" | "reject"

/**
 * Section-specific feedback for revision
 */
export interface SectionFeedback {
  /** Section identifier (e.g., "strategic_insights", "emerging_trends") */
  section: string
  /** Specific feedback for this section */
  feedback: string
  /** Index within the section array (if applicable) */
  index?: number
}

/**
 * Request to submit a digest review
 */
export interface DigestReviewRequest {
  /** Review action */
  action: ReviewAction
  /** Name of the reviewer */
  reviewer: string
  /** Section-specific feedback (key = section type, value = feedback) */
  section_feedback?: Record<string, string>
  /** General notes about the review */
  notes?: string
}

/**
 * Request to revise a specific section
 */
export interface ReviseDigestSectionRequest {
  /** Revision instructions */
  feedback: string
}

/**
 * Sort order for table sorting
 */
export type SortOrder = "asc" | "desc"

/**
 * Filters for digest list queries
 */
export interface DigestFilters {
  /** Filter by digest type */
  digest_type?: DigestType
  /** Filter by status */
  status?: DigestStatus
  /** Pagination limit */
  limit?: number
  /** Pagination offset */
  offset?: number
  /** Field to sort by */
  sort_by?: string
  /** Sort direction */
  sort_order?: SortOrder
}

/**
 * Progress event for digest generation (SSE)
 */
export interface DigestGenerationProgress {
  /** Task identifier */
  task_id: string
  /** Current step */
  step: string
  /** Progress percentage (0-100) */
  progress: number
  /** Status */
  status: "processing" | "completed" | "error"
  /** Error message if failed */
  error_message?: string
  /** Digest ID once created */
  digest_id?: number
}
